src/ethereum_spec_tools/sync.py

```diff
@@ -128,239 +71,93 @@ def next_fork(self) -> Optional[Hardfork]:
 
-    def make_header(self, json: Any) -> Any:
+    def take_block(self) -> Optional[Any]:
         """
-        Create a Header object from JSON describing it.
+        Pop a block of the download queue.
         """
-        return self.module("eth_types").Header(
-            hex_to_bytes32(json["parentHash"]),
-            hex_to_bytes32(json["sha3Uncles"]),
-            self.module("utils.hexadecimal").hex_to_address(json["miner"]),
-            hex_to_bytes32(json["stateRoot"]),
-            hex_to_bytes32(json["transactionsRoot"]),
-            hex_to_bytes32(json["receiptsRoot"]),
-            Bytes256(hex_to_bytes(json["logsBloom"])),
-            hex_to_uint(json["difficulty"]),
-            hex_to_uint(json["number"]),
-            hex_to_uint(json["gasLimit"]),
-            hex_to_uint(json["gasUsed"]),
-            hex_to_u256(json["timestamp"]),
-            hex_to_bytes(json["extraData"]),
-            hex_to_bytes32(json["mixHash"]),
-            hex_to_bytes8(json["nonce"]),
-        )
+        # Use a loop+timeout so that KeyboardInterrupt is still raised.
+        while True:
+            try:
+                return self.queue.get(timeout=1)
+            except Empty:
+                pass
 
-    def fetch_ommers(self, ommers_needed: Dict[int, int]) -> Dict[int, Any]:
+    def download(self) -> None:
         """
-        Fetch the ommers for a given block from the RPC provider.
+        Fetch chunks of blocks from the RPC provider.
         """
-        calls = []
-
-        for (block_number, num_ommers) in ommers_needed.items():
-            for i in range(num_ommers):
-                calls.append(
-                    {
-                        "jsonrpc": "2.0",
-                        "id": hex(block_number * 20 + i),
-                        "method": "eth_getUncleByBlockNumberAndIndex",
-                        "params": [hex(block_number), hex(i)],
-                    }
-                )
-
-        if calls == []:
-            return {}
-
-        data = json.dumps(calls).encode("utf-8")
-
-        self.log.debug(
-            "fetching ommers [%d, %d]...",
-            min(ommers_needed),
-            max(ommers_needed),
-        )
-
-        post = request.Request(
-            self.options.rpc_url,
-            data=data,
-            headers={
-                "Content-Length": str(len(data)),
-                "Content-Type": "application/json",
-            },
-        )
+        running = True
 
-        with request.urlopen(post) as response:
-            replies = json.load(response)
-            ommers: Dict[int, Dict[int, Any]] = {}
+        while running:
+            count = max(1, self.queue.maxsize // 2)
+            if self.next_fork:
+                # Don't fetch a batch across a fork boundary
+                count = min(count, self.next_fork.block - self.block_number)
+            replies = self.fetch_blocks(self.block_number, count)
```

Comment:
Am I correct in assuming that we can't batch across a fork boundary because something in the fetch code needs to understand the header? Something feels weird about having to split batches because of that.

Reply:
Fork A blocks need to processed by the code in `src/ethereum/A` and fork B blocks need to be processed by the code in `src/ethereum/B`. The current code tracks forks globally rather than on a per block basis. This means we must process all fork A blocks, switch to fork B and then start processing fork B blocks.

Up until now we had been getting away with this due to python duck typing.

We could track forks on a per block basis instead, but I can't see any benefit to doing that.